### PR TITLE
🔒 #1333 Lock features according to the pricing plan

### PR DESCRIPTION
**Pull request by stephanevieira75**
## Todo
- [x] set message history limit
- [x] add banner to the message feed
- [x] Take a look at feature flipping in React

### Locked guests feature popup
Capture d’écran de 2021-06-16 14-49-50

### Message limit banner
Capture d’écran de 2021-06-10 16-09-10

#### Member or guest message
Capture d’écran de 2021-06-21 16-46-55
#### Owner or admin message
Capture d’écran de 2021-06-21 16-46-34

### Blocked Company popup
Capture d’écran de 2021-06-21 16-47-30

### CODE REVIEW

**Comment by RomaricMourgues** [comment]
Hey! Text seems wrong to me, the 10k limit apply for messages in all Twake (not 10k per channel) just replace "channel" by "company" :)

File: twake/frontend/src/app/components/LockedFeaturesComponents/LockedGuestsPopup/LockedGuestsPopup.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import ObjectModal from 'app/components/ObjectModal/ObjectModal';
+import React from 'react';
+import ModalManager from 'app/components/Modal/ModalManager';
+import Languages from 'services/languages/languages.js';
+import RouterService from 'app/services/RouterService';
+import Emojione from 'app/components/Emojione/Emojione';
+
+const LockedGuestsPopup = (): JSX.Element => {
+  const { companyId } = RouterService.getStateFromRoute();
+  const pricingPlanUrl = `https://console.twake.app/companies/${companyId}/subscription`;
```
Comment (chamerling) [inline twake/frontend/src/app/components/LockedFeaturesComponents/LockedGuestsPopup/LockedGuestsPopup.tsx:11]: This should come from config so from props

File: twake/frontend/src/app/services/languages/locale/en.js [hunk 2]
```diff
@@ -1957,4 +1957,15 @@ export default {
   'general.user.role.company.guest': 'Guest',
   'general.user.deleted': 'Deleted Account',
   'scenes.app.channelsbar.channel_menu.guest_management': 'Guest management',
+  'components.locked_features.locked_history_banner.title':
+    'You have just reached the message limit',
+  'components.locked_features.locked_history_banner.description':
+    "Your channel has over 10000 messages. That's a lot! Upgrade now to access all your message history.",
+  'components.locked_features.locked_history_banner.button': 'Learn more',
+  'components.locked_features.locked_guests_popup.title':
+    'You are do not have access to guest features',
```
Comment (chamerling) [inline twake/frontend/src/app/services/languages/locale/en.js:1966]: `You do not have access to guests feature`

**Review by chamerling — CHANGES_REQUESTED** [review]

File: twake/frontend/src/app/components/LockedFeaturesComponents/LockedGuestsPopup/LockedGuestsPopup.tsx (hunk 1 above)
Comment (stephanevieira75) [reply twake/frontend/src/app/components/LockedFeaturesComponents/LockedGuestsPopup/LockedGuestsPopup.tsx]: Done 👍🏻

**Review by chamerling — COMMENTED** [review]
Looks like we do not have the same prettier configuration :)
Anyway, this is the way to implement this feature, good job!

**Review by chamerling — APPROVED** [review]
Looks good!

File: twake/frontend/src/app/scenes/Client/MainView/MainView.tsx [hunk 3]
```diff
@@ -36,6 +37,13 @@ const MainView: FC = () => {
       {!!channelId && ready && (
         <>
           <AccountStatusComponent />
+          {companyId && (
+            <CompanyBillingBanner
+              companyId={companyId}
+              // TODO ADD URL
+              url=""
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/MainView/MainView.tsx:44]: This must not be a props, CompanyBillingBanner can get it from InitService.server_infos....

File: twake/frontend/src/app/components/OnBoarding/CompanyBillingBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+export default ({ url, companyId }: PropsType): JSX.Element => {
+  const { Link } = Typography;
+  const userGroups: { [key: string]: CompanyType } = Groups.user_groups;
+  const { role, plan } = userGroups[companyId];
+  const isCurrentUserAdminOrOwner = role === 'admin' || role === 'owner';
+  const billingStatusError = plan?.billing?.status === 'error';
+
+  const onClickLink = () => window.open(url, 'blank');
+
+  const shouldDisplayBanner = isCurrentUserAdminOrOwner && billingStatusError;
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/components/OnBoarding/CompanyBillingBanner.tsx:21]: This code is great  but you must change it a bit according to issue from Benoit (members should also see the banner but with another text)

**Review by RomaricMourgues — CHANGES_REQUESTED** [review]

File: twake/frontend/src/app/components/OnBoarding/CompanyBillingBanner.tsx (hunk 4 above)
Comment (stephanevieira75) [reply twake/frontend/src/app/components/OnBoarding/CompanyBillingBanner.tsx]: Done !

File: twake/frontend/src/app/scenes/Client/MainView/MainView.tsx (hunk 3 above)
Comment (stephanevieira75) [reply twake/frontend/src/app/scenes/Client/MainView/MainView.tsx]: removed

File: twake/frontend/src/app/services/languages/locale/en.js [hunk 5]
```diff
@@ -1972,4 +1972,31 @@ export default {
   'components.richtexteditor.toolbar.ordered-list': 'Ordered list',
   'components.richtexteditor.toolbar.blockquote': 'Blockquote',
   'components.richtexteditor.toolbar.code-block': 'Code Block',
+  'components.locked_features.locked_history_banner.title':
+    'You have just reached the message limit',
+  'components.locked_features.locked_history_banner.description':
+    "Your company has over 10000 messages. That's a lot! Upgrade now to access all your message history.",
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/services/languages/locale/en.js:1978]: Change it to "Your company has a lot of messages! Upgrade now to access all your message history." (also the french version)

Comment (RomaricMourgues) [reply twake/frontend/src/app/services/languages/locale/en.js:1978]: @stephanevieira75

File: twake/frontend/src/app/services/Apps/Messages/MessageHistoryService.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+class MessageHistoryService {
+  shouldLimitMessages(firstMessageId: string): boolean {
+    const MONTH = 1000 * 60 * 60 * 24 * 30;
+    const DELAY = 3 * MONTH;
+
+    const firstMessageTimestamp = Numbers.timeuuidToDate(firstMessageId) * 1000;
+    const isFirstMessageOlderOrEqualThanDelay = firstMessageTimestamp >= DELAY;
+    const isActiveFeature = FeatureTogglesService.isActiveFeatureName(FeatureNames.MESSAGE_HISTORY);
+    return isFirstMessageOlderOrEqualThanDelay && !isActiveFeature;
+  }
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/services/Apps/Messages/MessageHistoryService.ts:14]: shouldLimitMessages should return true only when at least N (for instance 100) messages are loaded in the channel. You must add one more parameter to the function.

Comment (RomaricMourgues) [reply twake/frontend/src/app/services/Apps/Messages/MessageHistoryService.ts:14]: @stephanevieira75

File: twake/frontend/src/app/services/languages/locale/en.js (hunk 5 above)
Comment (stephanevieira75) [reply twake/frontend/src/app/services/languages/locale/en.js]: Done

File: twake/frontend/src/app/services/Apps/Messages/MessageHistoryService.ts (hunk 6 above)
Comment (stephanevieira75) [reply twake/frontend/src/app/services/Apps/Messages/MessageHistoryService.ts:15]: Done

**Review by chamerling — APPROVED** [review]

**Pull request merged**